[EIP-3074] Move to two-opcode version (#2)
quilt/EIPs: EIPS/eip-3074.md

---
eip: 3074

> title: AUTHORIZE and AUTHORIZEDCALL opcodes

Reviewer: ```suggestion
title: AUTHORIZEDCALL opcode
```

I prefer this, because it is the "meat" of the EIP. `AUTHORIZE` is really just a helper op.

Author: I think I prefer "AUTH and AUTHCALL opcodes", because those might grow to a family of opcodes in the future

## Simple Summary

> Creates two new EVM instructions that allow an externally owned account (EOA) to authorize a specific contract via an ECDSA signature to send calls on it's behalf.

Reviewer 2: `s/it's/its/`

Also, I was told to keep the simple summary as simple as possible. Maybe:

```suggestion
Creates two new EVM instructions that authorize (via an ECDSA signature) a contract to act on behalf of an externally owned account.
```

## Abstract

> This EIP creates two new EVM instruction. The first, `AUTHORIZE`,  sets a context variable `authorizedAccount` based on an ECDSA signature. The second, `AUTHORIZEDCALL`, sends a call on behalf of `authorizedAccount`,  setting the caller address to and deducting the call's value from the authorized account.

Reviewer: ```suggestion
This EIP introduces two EVM instruction `AUTHORIZE` and `AUTHORIZEDCALL`. The first sets a context variable `authorizedAccount` based on an ECDSA signature. The second sends a call as the `authorizedAccount`.
```

## Specification

### Conventions

 - **`top - N`** - the `N`th most recently pushed value on the EVM stack, where `top - 0` is the most recent.
 - **`||`** - byte concatenation operator.

### Constants

|     Constant     | Value  |
| ---------------- | ------ |
| `TYPE`           | `0x03` |

`TYPE` is an [EIP-2718](./eip-2718.md) transaction type reserved for EIP-3074 signatures to prevent signature collisions.

### Context Variables

|     Variable        | Type      | Initial Value                                                                        |
| ------------------- | --------- |:------------------------------------------------------------------------------------ |

> | `authorizedAccount` | `address` | `0x0000000000000000000000000000000000000000` |

Reviewer: Does `0x0000000000000000000000000000000000000000 == 0000000000000000000000000000000000000000000000000000000000000000`? Because we wouldn't want a change in address format to break things.

Author: I think one alternative would be to have it be an optional address (or address pointer), so that we don't have to use the zero address to indicate no authorized account

Reviewer 2: Once pushed on the stack, addresses are extended to 32 bytes, right?

Reviewer: @SamWilsn yes, I'm just trying to think if there is any weirdness that could occur.

> `authorizedAccount` indicates the address of an account the executing contract is authorized to send calls from. The zero address indicates that no such account is set. The variable has the same scope as the program counter: It persists throughout execution of the contract, but is not passed along any calls (including `DELEGATECALL`).

Reviewer: ```suggestion
`authorizedAccount` indicates the address of an account the executing contract is authorized to send calls from. The zero address indicates that no such account is set. The variable has the same scope as the program counter -- it persists throughout execution of the contract, but is not passed along any calls (including `DELEGATECALL`).
```

### `AUTHORIZE` (`0xf8`)

A new opcode `AUTHORIZE` is created at `0xf8`. It takes in 4 stack elements and returns 1 stack element.

#### Input

| Stack      | Value        |
| ---------- | ------------ |
| `top - 0`  | `commit`     |
| `top - 1`  | `yParity`    |
| `top - 2`  | `r`          |
| `top - 3`  | `s`          |

#### Output

| Stack      | Value               |
| ---------- | ------------------- |
| `top - 0`  | `authorizedAccount` |

#### Behavior

> The arguments (`yParity`, `r`, `s`) are interpreted as an ECDSA signature of the form `secp256k1(keccak256(TYPE || abi.encode(invoker, commit)))`, where:

Reviewer: ```suggestion
The arguments (`yParity`, `r`, `s`) are interpreted as an ECDSA signature on the secp256k1 curve over the message `keccak256(TYPE || rlp([invoker, commit]))`, where:
```

Reviewer: At minimum, I'd like to use rlp(...) instead of abi.encode(...).

Reviewer: or by hand >:(

Author: Changed it to specifying it by hand.

 - `invoker` is the address of the contract executing `AUTHORIZE`.

>  - `commit` is one of the arguments passed into `AUTHORIZE` and is a 32-byte value that can be used to commit to specific additional validity condition in the invoker's pre-processing logic.

Reviewer: ```suggestion
 - `commit` is one of the arguments passed into `AUTHORIZE` and is a 32-byte value that can be used to commit to specific validity condition (e.g. nonce) in the invoker's pre-processing logic.
```

If the signature is valid, the `signerAddress` is recovered. If `signerAddress != tx.origin`, the context variable `authorizedAccount` is set to `signerAddress`.

> In any other case, i.e. if the signature is invalid or `signerAddress == tx.origin`, `authorizedAccount` is reset to the zero address.

Reviewer: Would rather have a bullet point list that explains what an "invalid signature" is.

> `AUTHORIZE` returns the stack element `authorizedAccount`, which is set to the context variable of the same name.

Reviewer: ```suggestion
`AUTHORIZE` returns the stack element `authorizedAccount`, which is set to the recovered address.
```

#### Gas Cost

> The gas cost for `AUTHORIZE` is `3000`, which is the same cost as for the existing ecrecover precompile.

Reviewer: ```suggestion
The gas cost for `AUTHORIZE` is `3000`. This is the same cost as for the `ecrecover` precompile.
```

### `AUTHORIZEDCALL` (`0xf9`)

> A new opcode `AUTHORIZEDCALL` is created at `0xf9`. It takes in 7 stack elements and returns 1 stack element. It closely mirrors the behavior of the existing `CALL` (`0xF1`) instruction.

Reviewer: ```suggestion
A new opcode `AUTHORIZEDCALL` is created at `0xf9`. It takes in 7 stack elements and returns 1 stack element. It matches the behavior of the existing `CALL` (`0xF1`) instruction, except where noted below.
```

#### Input

| Stack     | Value        |
| --------- | ------------ |
| `top - 0` | `gas`        |

> | `top - 1` | `callee`     |

Reviewer: ```suggestion
| `top - 1` | `addr`     |
```

This is the common "gav" notation for calls.

| `top - 2` | `value`      |
| `top - 3` | `argsOffset` |
| `top - 4` | `argsLength` |
| `top - 5` | `retOffset`  |
| `top - 6` | `retLength`  |

#### Output

| Stack      | Value     |
| ---------- | --------- |
| `top - 0`  | `success` |

#### Behavior

> `AUTHORIZEDCALL` has the same execution logic as `CALL`, except for:

Reviewer: ```suggestion
`AUTHORIZEDCALL` is interpreted the same as `CALL`, except for:
```

>  - If `authorizedAccount` is the zero address, no call is made and `success` is set to `0`. Otherwise, the caller address for the call is set to `authorizedAccount`.

Reviewer: Can we separate this into two bullet points?

>  - The call value is deducted from the `authorizedAccount` balance. If that balance is insufficient, no call is made and `success` is set to `0`.

Reviewer: I don't think this should be in the EIP. It is opening up a can of worms that I don't think we should open right now. Having the invoker maintain a balance will be good enough for 90% of cases.

> `AUTHORIZEDCALL` must increase the call depth by one, in the same way as `CALL`. `AUTHORIZEDCALL` must not increase the call depth by two (as it would if it first called into the authorized account and then into the callee.)

Reviewer: ```suggestion
`AUTHORIZEDCALL` must increase the call depth by one. `AUTHORIZEDCALL` must not increase the call depth by two as if it first called into the authorized account and then into the target.
```

The return data area accessed with `RETURNDATASIZE` (`0x3d`) and `RETURNDATACOPY` (`0x3e`) must be set in the same way as the `CALL` instruction.

> In a static context (such as the one created by `STATICCALL`), `AUTHORIZEDCALL` with a non-zero `value` must exit the current execution frame immediately (in the same way `CALL` behaves with a non-zero value in a static context.)

Reviewer: I don't think this shouldn't be specified since it is inheriting this property from `CALL`.

#### Gas Cost

`AUTHORIZEDCALL` has the same gas cost as `CALL`, except for:

>  - The dynamic gas cost for non-zero call value is increased from `9000` to `21000`.

Reviewer: See above.

## Specification

### Conventions

 - **`top - N`** - the `N`th most recently pushed value on the EVM stack, where `top - 0` is the most recent.
 - **`||`** - byte concatenation operator.

### Constants

|     Constant     | Value  |
| ---------------- | ------ |
| `TYPE`           | `0x03` |

`TYPE` is an [EIP-2718](./eip-2718.md) transaction type reserved for EIP-3074 signatures to prevent signature collisions.

### Context Variables

|     Variable        | Type      | Initial Value |
| ------------------- | --------- |:------------- |
| `authorizedAccount` | `address` | unset         |

> `authorizedAccount` indicates the address of an account the executing contract is authorized to send calls from. An unset value indicates that no such account is set. The variable has the same scope as the program counter -- it persists throughout execution of the contract, but is not passed along any calls (including `DELEGATECALL`).

Reviewer 2: ```suggestion
The context variable `authorizedAccount` shall indicate the active account for `AUTHCALL` instructions in the current frame of execution. If set, `authorizedAccount` shall only contain an account which has given the contract authorization to act on its behalf. An unset value shall indicate that no such account is set, and that there is not yet an active account for `AUTHCALL` instructions in the current frame of execution.

The variable has the same scope as the program counter -- `authorizedAccount` persists throughout a single frame of execution of the contract, but is not passed through any calls (including `DELEGATECALL`). If the same contract is being executed in separate execution frames (ex. a `CALL` to self), both frames shall have independent values for `authorizedAccount`. Initially in each frame of execution, `authorizedAccount` is always unset, even if a previous execution frame for the same contract has a value.
```

### `AUTH` (`0xf6`)

> A new opcode `AUTH` is created at `0xf6`. It takes in 4 stack elements and returns 1 stack element.

Reviewer 2: ```suggestion
A new opcode `AUTH` shall be created at `0xf6`. It shall take four stack element inputs and returns one stack element.
```

#### Input

| Stack      | Value        |
| ---------- | ------------ |
| `top - 0`  | `commit`     |
| `top - 1`  | `yParity`    |
| `top - 2`  | `r`          |
| `top - 3`  | `s`          |

#### Output

| Stack      | Value               |
| ---------- | ------------------- |
| `top - 0`  | `authorizedAccount` |

#### Behavior

The arguments (`yParity`, `r`, `s`) are interpreted as an ECDSA signature on the secp256k1 curve over the message `keccak256(TYPE || paddedInvokerAddress || commit)`, where:

>  - `paddedInvokerAddress` is the 32-byte right-alligned address of the contract executing `AUTH`, padded with zero-bytes.

Reviewer 2: ```suggestion
 - `paddedInvokerAddress` is the address of the contract executing `AUTH`, left-padded with zeroes to a total of 32 bytes (ex. `0x000000000000000000000000AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA`).
```

>  - `commit` is one of the arguments passed into `AUTH` and is a 32-byte value that can be used to commit to specific additional validity condition in the invoker's pre-processing logic (e.g. a nonce for replay protection).

Reviewer 2: ```suggestion
 - `commit`, one of the arguments passed into `AUTH`, is a 32-byte value that can be used to commit to specific additional validity conditions in the invoker's pre-processing logic (e.g. a nonce for replay protection).
```

If the signature is valid, the `signerAddress` is recovered. Signature validity and signer recovery is handled analogous to transaction signatures, including the stricter `s` range for preventing ECDSA malleability. Note that `yParity` is expected to be `0` or `1`.  If `signerAddress != tx.origin`, the context variable `authorizedAccount` is set to `signerAddress`.
In any other case, i.e. if the signature is invalid or `signerAddress == tx.origin`, `authorizedAccount` is reset to an unset value.

`AUTH` returns the new `authorizedAccount` if set, or `0` otherwise.

#### Gas Cost

> The gas cost for `AUTH` is `3000`. This is the same cost as for the `ecrecover` precompile.

Reviewer 2: Does calling into the `ecrecover` precompile cost 3000 gas total, or is it 3000 plus the cost of a call?

If it's 3000+call, I think this needs to be a bit more expensive.

Author: Looked into this. Turns out the full cost of an ecrecover (via `STATICCALL`) after Berlin will just be this 3000 gas plus a single `WARM_STORAGE_READ_COST` (100) as per EIP-2929. `AUTH` has none of the static call overhead. Both `ORIGIN` and `CALLER` have a gas cost of 2, so I think any extra cost here should also be on the order of ~10 gas at most - at which point a clean 3000 might just be preferable...

Author: Ah okay, I did actually find one extra step that is somewhat expensive, namely the hashing for the signed message. That would natively have a gas cost of 48. So I feel like the realistic values for `AUTH` gas cost could be 3000 / 3050 / 3100.

### `AUTHCALL` (`0xf7`)

> A new opcode `AUTHCALL` is created at `0xf7`. It takes in 7 stack elements and returns 1 stack element. It matches the behavior of the existing `CALL` (`0xF1`) instruction, except where noted below.

Reviewer 2: ```suggestion
A new opcode `AUTHCALL` shall be created at `0xf7`. It shall take seven stack elements and return one stack element. It matches the behavior of the existing `CALL` (`0xF1`) instruction, except where noted below.
```

#### Input

| Stack     | Value        |
| --------- | ------------ |
| `top - 0` | `gas`        |
| `top - 1` | `addr`     |
| `top - 2` | `value`      |
| `top - 3` | `argsOffset` |
| `top - 4` | `argsLength` |
| `top - 5` | `retOffset`  |
| `top - 6` | `retLength`  |

#### Output

| Stack      | Value     |
| ---------- | --------- |
| `top - 0`  | `success` |

#### Behavior

`AUTHCALL` is interpreted the same as `CALL`, except for:

>  - If `authorizedAccount` is unset, execution is considered invalid and must exit the current execution frame immediately (in the same way it would do e.g. for a stack underflow or an invalid jump).

Reviewer 2: ```suggestion
 - If `authorizedAccount` is unset, execution is considered invalid and must exit the current execution frame immediately (in the same way as a stack underflow or invalid jump).
```

 - Otherwise, the caller address for the call is set to `authorizedAccount`.

The call value is deducted from the balance of the executing contract. It is not paid by the `authorizedAccount`.

`AUTHCALL` must increase the call depth by one. `AUTHCALL` must not increase the call depth by two as if it first called into the authorized account and then into the target.

The return data area accessed with `RETURNDATASIZE` (`0x3d`) and `RETURNDATACOPY` (`0x3e`) must be set in the same way as the `CALL` instruction.

Importantly, `AUTHCALL` does not reset `authorizedAccount`, but leaves it unchanged.

#### Gas Cost

`AUTHCALL` has the same gas cost as `CALL`.

As with `CALL`, the gas cost for the opcode itself (both the static and the dynamic portion) is always charged, independent of whether the call is actually executed. The gas passed into the call is calculated following EIP-150 and is refunded partially if the call returns with unused gas left, or completely if the call is not executed at all because of a failing pre-check.

## Rationale

### Throwing for Unset `authorizedAccount` During `AUTHCALL`

> Any correcly behaving contract should never reach an `AUTHCALL` without having successfully set `authorizedAccount` before. The correct behavior is thus to exit the current execution frame immediately. This is especially important in the context of transaction sponsoring / relaying, which is expected to be one of the main use cases for this EIP. There, being unable to distinguish between a sponsee-attributable failure (like a failing sub-call) and a sponsor-attributable failure (like a failing `AUTH`) is especially dangerous and thus must be prevented.

Reviewer 2: ```suggestion
A well-behaved contract should never reach an `AUTHCALL` without having successfully set `authorizedAccount` beforehand. The safest behavior, therefore, is to exit the current frame of execution immediately. This is especially important in the context of transaction sponsoring / relaying, which is expected to be one of the main use cases for this EIP. In a sponsored transaction, the inability to distinguish between a sponsee-attributable fault (like a failing sub-call) and a sponsor-attributable fault (like a failing `AUTH`) is especially dangerous and should be prevented because it charges unfair fees to the sponsee.
```
